File: nxplx-backend/NxPlx/TMdbApi.fs
module NxPlx.TMdbApi

open FSharp.Control.Tasks
open NxPlx.Types
open System.Net.Http
open Newtonsoft.Json
open System
open System.IO
open System.Web
open LiteDB
open LiteDB.FSharp.Extensions

type TMdbFilmResult = {vote_count:int; id:int; video:bool; vote_average:float32; title:string; popularity:float32;
                       poster_path:string; original_language:string; original_title:string; genre_ids: int array;
                       backdrop_path:string; adult:bool; overview:string; release_date:DateTime}

type TMdbSeriesResult = {original_name:string; id:int; name:string; vote_count:int; vote_average:float32;
                         poster_path:string; first_air_date:DateTime; popularity:float32; genre_ids:int array;
                         original_language:string; backdrop_path:string; overview:string; origin_country:string array}


type TMdbEpisodeCrewResult = {id:int; credit_id:string; name:string; department:string; job:string; profile_path:string}
type TMdbEpisodeGuestStarResult = {id:int; name:string; credit_id:string; character:string;
                                   order:int; gender:int; profile_path:string}
type TMdbEpisodeResult = {air_date:DateTime; episode_number:int; id:int; name:string; overview:string;
                          production_code:string; season_number:int; show_id:int; still_path:string;
                          vote_average:float32; vote_count:int; crew:TMdbEpisodeCrewResult array;
                          guest_stars:TMdbEpisodeGuestStarResult array}


type TMdbFilmSearch = {page:int; total_results:int; total_pages:int; results: TMdbFilmResult array}
type TMdbSeriesSearch = {page:int; total_results:int; total_pages:int; results: TMdbSeriesResult array}
type TMdbSeasonSearch = {_id:string; air_date:DateTime; episodes:TMdbEpisodeResult array; name:string;
                          overview:string; id:int; poster_path:string; season_number:int}

[<CLIMutable>]
type CachedTMdbSearchRequest = {id:string; content:string; expires:DateTime; }

     
let fetchWeb (httpClient:HttpClient) (url:string) = async {
    let! response = httpClient.GetAsync url |> Async.AwaitTask
    let! content = response.Content.ReadAsStringAsync() |> Async.AwaitTask
    return content
}    
    
let fetch (cache:LiteCollection<CachedTMdbSearchRequest>) (httpClient:HttpClient) (validTimeSpan:TimeSpan) (url:string) =
    let expiryDate = DateTime.UtcNow.Add validTimeSpan
    
    let bsonId = BsonValue url;
    let cachedOption = cache.TryFindById bsonId
    
    match cachedOption with
    | Some c ->
        if c.expires > DateTime.UtcNow  then
            c.content
        else
            let content = fetchWeb httpClient url |> Async.RunSynchronously
            cache.Update {id=url; content=content; expires=expiryDate} |> ignore
            content
    | None ->
        let content = fetchWeb httpClient url |> Async.RunSynchronously
        cache.Insert {id=url; content=content; expires=expiryDate} |> ignore
        content

type TMdbApiKey = ApiKeyValue of string | ApiKeyPath of string

type TMdbApi(apiKey:TMdbApiKey) =
    
    let key = match apiKey with
              | ApiKeyValue k -> k
              | ApiKeyPath kp -> System.IO.File.ReadAllText kp
    
    let db =
        let dir = Path.Combine("data", "database");
        Directory.CreateDirectory dir |> ignore
        let db = new LiteDatabase(Path.Combine(dir, "tmdb-cache.litedb"))
        db.GetCollection<CachedTMdbSearchRequest> "requests"
       
    
    let client =
        let client = new HttpClient()
        client.DefaultRequestHeaders.Add("User-Agent", "NxPlx")
        client
       
    let _fetchAndDeserialize validTimeSpan url : 'a =
        fetch db client validTimeSpan url |> JsonConvert.DeserializeObject<'a>
        
    let _fetch validTimeSpan url =
        fetch db client validTimeSpan url
        
    member __.downloadImage (size:string) (posterUrl:string) =
        let dir = Path.Combine ("data", "posters")
        let run = task {
            Directory.CreateDirectory dir |> ignore
            let posterName = Path.GetFileNameWithoutExtension posterUrl
            let outputPath = Path.Combine (dir, (sprintf "%s-%s.jpg" posterName size))
            if File.Exists outputPath then return outputPath
            else 
                let! response = client.GetAsync((sprintf "http://image.tmdb.org/t/p/%s%s?api_key=%s" size posterUrl key))
                use! inputStream = response.Content.ReadAsStreamAsync()
                use outputStream = File.OpenWrite outputPath
                do! inputStream.CopyToAsync outputStream
                return outputPath
        }
        run |> ignore
    member __.getFilmDetails id =    
        _fetch (TimeSpan.FromDays 500.0) (sprintf "https://api.themoviedb.org/3/movie/%s?api_key=%s" id key)
    member __.getSeriesDetails id =    
        _fetch (TimeSpan.FromDays 50.0) (sprintf "https://api.themoviedb.org/3/tv/%s?api_key=%s" id key)
    member __.getEpisodeDetails id season episode =    
        _fetch (TimeSpan.FromDays 500.0) (sprintf "https://api.themoviedb.org/3/tv/%s/season/%s/episode/%s?api_key=%s" id season episode key)
        
    
    member __.findFilm (film:FilmEntry) : TMdbFilmSearch  =
        let year = match film.year with | -1 -> "" | _ -> sprintf "&year=%i" film.year        
        _fetchAndDeserialize (TimeSpan.FromDays 500.0) (sprintf "https://api.themoviedb.org/3/search/movie?api_key=%s&query=%s%s" key (HttpUtility.UrlEncode film.title) year)
    member __.findSeries (name:string) : TMdbSeriesSearch =
        _fetchAndDeserialize (TimeSpan.FromDays 500.0) (sprintf "https://api.themoviedb.org/3/search/tv?api_key=%s&query=%s" key (HttpUtility.UrlEncode name))
    member __.findSeason  id season : TMdbSeasonSearch   =
        _fetchAndDeserialize (TimeSpan.FromDays 100.0) (sprintf "https://api.themoviedb.org/3/tv/%i/season/%i?api_key=%s" id season key)

let tmdbApi = new TMdbApi(ApiKeyPath(Path.Combine("data", "tmdb-api-key")))

File: nxplx-backend/NxPlx.Integrations.TMDB/TmdbApi.cs
using System;
using System.IO;
using System.Net.Http;
using System.Threading.Tasks;
using System.Web;
using Newtonsoft.Json;
using NxPlx.Abstractions;
using NxPlx.Infrastructure.IoC;
using NxPlx.Integrations.TMDBApi.Models.Movie;
using NxPlx.Integrations.TMDBApi.Models.Search;
using NxPlx.Integrations.TMDBApi.Models.Tv;
using NxPlx.Integrations.TMDBApi.Models.TvSeason;
using NxPlx.Services.Caching;

namespace NxPlx.Integrations.TMDBApi
{

    public interface IDetailsApi
    {
        
    }
    
    public class TmdbApi
    {
        private const string ImageFolder = "images";

        private const string BaseUrl = "https://api.themoviedb.org/3";
        
        private string _key;
        private ICachingService _cachingService;
        private HttpClient _client = new HttpClient
        {
            DefaultRequestHeaders =
            {
                {"User-Agent", "NxPlx"}
            }
        };


        public TmdbApi(string apiKey)
        {
            _key = apiKey;
            
            var container = new ResolveContainer();
            _cachingService = container.Resolve<ICachingService>();
            Directory.CreateDirectory(Path.Combine("data", ImageFolder));
        }
        

        public async Task<SearchResult<MovieResult>> SearchMovies(string title, int year)
        {
            var encodedTitle = HttpUtility.UrlEncode(title);
            var encodedYear = year == 0 ? "" : $"&year={year}";
            var url = $"{BaseUrl}/search/movie?api_key={_key}&query={encodedTitle}{encodedYear}";

            var content = await Fetch(url);
            return JsonConvert.DeserializeObject<SearchResult<MovieResult>>(content);
        }
        
        public async Task<SearchResult<TvShowResult>> SearchTvShows(string name)
        {
            var encodedTitle = HttpUtility.UrlEncode(name);
            var url = $"{BaseUrl}/search/tv?api_key={_key}&query={encodedTitle}";

            var content = await Fetch(url);
            return JsonConvert.DeserializeObject<SearchResult<TvShowResult>>(content);
        }

        public async Task<MovieDetails> FetchMovieDetails(int id)
        {
            var url = $"{BaseUrl}/movie/{id}?api_key={_key}";
            
            var content = await Fetch(url);
            return JsonConvert.DeserializeObject<MovieDetails>(content);
        }
        
        public async Task<TvDetails> FetchTvDetails(int id)
        {
            var url = $"{BaseUrl}/tv/{id}?api_key={_key}";
            
            var content = await Fetch(url);
            return JsonConvert.DeserializeObject<TvDetails>(content);
        }
        
        public async Task<TvSeasonDetails> FetchTvSeasonDetails(int id, int season)
        {
            var url = $"{BaseUrl}/tv/{id}/season/{season}?api_key={_key}";

            var content = await Fetch(url);
            return JsonConvert.DeserializeObject<TvSeasonDetails>(content);
        }
        
        public async Task DownloadImage(string size, string imageUrl)
        {
            if (string.IsNullOrEmpty(imageUrl)) return;
            
            var imageName = Path.GetFileNameWithoutExtension(imageUrl.Trim('/'));
            var outputPath = Path.Combine(Path.Combine("data", ImageFolder, $"{imageName}-{size}.jpg"));

            if (!File.Exists(outputPath))
            {
                var response = await _client.GetAsync($"https://image.tmdb.org/t/p/{size}{imageUrl}?api_key={_key}");
                using (var imageStream = await response.Content.ReadAsStreamAsync())
                {
                    try
                    {
                        using (var outputStream = File.OpenWrite(outputPath))
                        {
                         await imageStream.CopyToAsync(outputStream);
                        }
                    }
                    catch (IOException e)
                    {
                        
                    }
                }
                
            }
        }
        
        private async Task<string> Fetch(string url)
        {
            var content = await _cachingService.GetAsync(url);

            if (string.IsNullOrEmpty(content))
            {
                var response = await _client.GetAsync(url);
                content = await response.Content.ReadAsStringAsync();
                
                await _cachingService.SetAsync(url, content, CacheKind.WebRequest);
            }

            return content;
        }

    }
}